File: 백준/Silver/2644. 촌수계산/촌수계산.py
import sys
from collections import deque

input = sys.stdin.readline

# 사람의 수
n = int(input())

a, b = map(int, input().split())

relations = [[] for _ in range(101)]
visited = [0 for _ in range(101)]

# 부모, 자식 관계의 수
m = int(input())

for _ in range(m):
    x, y = map(int, input().split())
    relations[x].append(y)
    relations[y].append(x)

def bfs(V):
    queue = deque([V])
    while queue:
        q = queue.popleft()
        for i in relations[q]:
            if not visited[i]:
                visited[i] = visited[q] + 1
                queue.append(i)

bfs(a)

print(visited[b] if visited[b] else -1)

File: 백준/Silver/2644. 촌수계산/촌수계산.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let lines = [];
let N, start, end, M;
let graph = [];
let visited = [];

rl.on('line', (line) => {
    lines.push(line);
});

rl.on('close', () => {
    N = parseInt(lines[0]);
    [start, end] = lines[1].split(' ').map(Number);
    M = parseInt(lines[2]);
    
    graph = Array.from({ length: N + 1 }, () => []);
    visited = Array(N + 1).fill(0);
    
    for (let i = 3; i < 3 + M; i++) {
        const [a, b] = lines[i].split(' ').map(Number);
        graph[a].push(b);
        graph[b].push(a);
    }
    
    dfs(graph, visited, start);
    
    console.log(visited[end] ? visited[end] : -1);
});

function dfs(graph, visited, v) {
    for (const j of graph[v]) {
        if (!visited[j]) {
            visited[j] = visited[v] + 1;
            dfs(graph, visited, j);
        }
    }
}